### ISSUE TITLE
Variant Ownership Transfer issue

### ISSUE THREAD

**Issue by nvanfleet** [migueldeicaza/SwiftGodot#102]
Not sure why I never saw this before but with this code it will crash when you try to assign value

```
        let gArray = GArray()
        gArray.pushBack(value: Variant("Hello"))
        let value = gArray[Int(0)]
        GD.print("\(value)")

```

```
        Thread 0 Crashed::  Dispatch queue: com.apple.main-thread
0   libsystem_kernel.dylib                 0x1951dc724 __pthread_kill + 8
1   libsystem_pthread.dylib                0x195213c28 pthread_kill + 288
2   libsystem_c.dylib                      0x195121ae8 abort + 180
3   libsystem_malloc.dylib                 0x195042e28 malloc_vreport + 908
4   libsystem_malloc.dylib                 0x1950595d4 malloc_zone_error + 104
5   libsystem_malloc.dylib                 0x195051148 nanov2_guard_corruption_detected + 44
6   libsystem_malloc.dylib                 0x195050344 nanov2_allocate_outlined + 404
7   libobjc.A.dylib                        0x194e77e00 class_createInstance + 64
8   CoreFoundation                         0x1952771c0 __CFAllocateObject + 24
9   CoreFoundation                         0x195290ca0 __NSArrayM_new + 60
10  AppKit                                 0x1985040a4 -[NSApplication _copyWindows] + 124
11  AppKit                                 0x19850400c -[NSApplication windows] + 160
12  AppKit                                 0x198671748 -[NSPersistentUIManager flushAllChanges] + 340
13  AppKit                                 0x198c72cf8 __59-[NSPersistentUIFlushScheduler _updateApplicationIsActive:]_block_invoke + 136
14  libdispatch.dylib                      0x195064400 _dispatch_client_callout + 20
15  libdispatch.dylib                      0x195067884 _dispatch_continuation_pop + 504
16  libdispatch.dylib                      0x19507ae7c _dispatch_source_invoke + 1588
17  libdispatch.dylib                      0x195072b4c _dispatch_main_queue_drain + 756
18  libdispatch.dylib                      0x195072848 _dispatch_main_queue_callback_4CF + 44
19  CoreFoundation                         0x195333c54 __CFRUNLOOP_IS_SERVICING_THE_MAIN_DISPATCH_QUEUE__ + 16
20  CoreFoundation                         0x1952f13d4 __CFRunLoopRun + 1992
21  CoreFoundation                         0x1952f04b8 CFRunLoopRunSpecific + 612
22  HIToolbox                              0x19eb3ac40 RunCurrentEventLoopInMode + 292
23  HIToolbox                              0x19eb3a8d0 ReceiveNextEventCommon + 220
24  HIToolbox                              0x19eb3a7d4 _BlockUntilNextEventMatchingListInModeWithFilter + 76
25  AppKit                                 0x198511d44 _DPSNextEvent + 636
26  AppKit                                 0x198510ee0 -[NSApplication(NSEvent) _nextEventMatchingEventMask:untilDate:inMode:dequeue:] + 716
27  Godot                                  0x1011eb9d8 0x100df8000 + 4143576
28  Godot                                  0x1011d24c0 0x100df8000 + 4039872
29  Godot                                  0x1011fbee4 0x100df8000 + 4210404
30  dyld                                   0x194ebbf28 start + 2236
```

**Comment by migueldeicaza** [comment]
My notes on what the problem is.

Address Sanitizer explains that what is happening here is the following:

1. When the method goes out of scope, the `deinit` for the Variant created in Swift is called, this calls the variant_destroy method in the Godot runtime.
2. The Array goes out of scope, and Swifth calls the `deinit` for it, this calls the Godot Array destructor, which in turn attempts to destroy the data from (1).

And variants do not seem to have reference counts, so it looks like this needs to be manually intercepted and handled specially.

Digging into the C# support, it looks like Array does *in fact* handle setting values by index specially, but not their append method [1].  And their "Variant" has a queue system to handle the disposing [2].  I do not understand why their append method does not need it yet.

Their C++ sample binding does not seem to do this, so I suspect they will have the same sort of crashes that Swift has right now.

But the support for manually handling Variants is pervasive (`Variant.CreateTakingOwnershipOfDisposableValue`, with their barely used companion `Variant.CreateCopyingBorrowed`).

I suspect that given this, that every use of Variant that goes *into* Godot will need to be duplicated.   Will check a few other bindings to see what they are doing.

Considering the above, I am removing the `deinit` destructor, which would only introduce a leak in cases where a variant is used purely in Swift-space, and never to be passed to Godot.   I think that this is a temporary workaround that should unblock you.


[1] modules/mono/glue/GodotSharp/GodotSharp/Core/Array.cs
[2] ./modules/mono/glue/GodotSharp/GodotSharp/Core/Variant.cs

**Comment by nvanfleet** [comment]
I was working on my project which seems to have some really out of control memory usage. Still trying to figure it out but I definitely wasn't sure exactly how memory management worked for Refcount'ed classes or for Nodes which you have to call QueueFree. I was pondering exactly what I should be doing in the Swift world, interesting to know that hypothetically it should be the same (things should be controlled by arc except for I guess Nodes?)

**Comment by migueldeicaza** [comment]
Would love if you could run your project with Instruments, so we could get an idea of which objects are being kept alive, and drill down from there.

There is a bit more to the story on references.

The Swift bindings are indeed keeping a ref-counted set of objects that point to the Godot world for anything class (anything deriving from Object, Wrapped, RefCounted, for example).   But this is merely a Swift-level set of reference counts, so they are great in the Swift world.

When the last Swift reference goes away, this triggers the destructor on the Godot/C++ side.   For a class of objects that just deletes the object, for anything that subclasses "Ref", it just drops the reference count by one, and if it was the last reference, it is removed.

**Comment by nvanfleet** [comment]
> Would love if you could run your project with Instruments, so we could get an idea of which objects are being kept alive, and drill down from there.
> 
> There is a bit more to the story on references.
> 
> The Swift bindings are indeed keeping a ref-counted set of objects that point to the Godot world for anything class (anything deriving from Object, Wrapped, RefCounted, for example). But this is merely a Swift-level set of reference counts, so they are great in the Swift world.
> 
> When the last Swift reference goes away, this triggers the destructor on the Godot/C++ side. For a class of objects that just deletes the object, for anything that subclasses "Ref", it just drops the reference count by one, and if it was the last reference, it is removed.

I guess I need to run it from SwiftGodotKit? I've currently not tried that, which definitely limits my debugging ability since it's not in an iOS application, but might be really helpful.

**Comment by migueldeicaza** [comment]
You do not need SwiftGodotKit, just launch Instruments, select the "Leaks" instrument, and you can either launch Godot, or attach to a running Godot and start recording.   Can be done on Mac or iOS.

**Comment by nvanfleet** [comment]
```
    private func testMeshInstanceLeak() {
        GD.print("going")
        self.displayed?.queueFree()
        
        let cylinder = CylinderMesh()
        cylinder.topRadius = 10
        cylinder.bottomRadius = 9
        cylinder.height = 40
        cylinder.radialSegments = 200
        
        let surfaceTool = SurfaceTool()
        surfaceTool.begin(primitive: Mesh.PrimitiveType.primitiveTriangles)
        surfaceTool.appendFrom(existing: cylinder, surface: 0, transform: Transform3D())
        surfaceTool.generateNormals()
        let array = surfaceTool.commit(existing: nil)
        surfaceTool.clear()

        let meshInstance = MeshInstance3D()
        meshInstance.mesh = array
        self.addChild(node: meshInstance)
        self.displayed = meshInstance
    }
```

So this code seems to just leak CylinderMesh and MeshInstance3D.

<img width="818" alt="Screenshot 2023-09-20 at 8 52 48 AM" src="https://github.com/migueldeicaza/SwiftGodot/assets/1486399/7bcde636-2164-4b52-a275-a141dbc58ff6">

**Comment by migueldeicaza** [comment]
Fixed
